### PR TITLE
Refactor loss ctx

### PR DESCRIPTION
**Pull request by jayhenry**
### Summary
This PR refactors the loss subsystem (`LossContext`) to standardize how loss contexts are constructed, sequence-parallelism (SP) is handled, and batches are built. 
Now Trainer/Worker only interact with LossConfig and LossContext objects.
It introduces unified builders across losses, removes legacy input-item types.

### Background
`LossContext` handles three things:
1. sequece parallel for inputs and outputs
2. [global loss calibration](https://xtuner.readthedocs.io/zh-cn/latest/pretrain_sft/advanced_tutorial/loss.html) for exact token-level mean loss regardless of dp size and grad accumalations 
3. chunk loss for less gpu memory consumption

### Highlights
- Introduces `BaseLossConfig.build`, and `BaseLossContext.build_batches`. Removes `*ContextInputItem`, and `BaseLossKwargs` now include built-in `to`, `sp_split` instead.
- CE loss:
  - Adds `CELossConfig.build` and `CELossContext.build_batches`.
- RL loss:
  - Adds `BaseRLLossConfig`, `BaseRLLossKwargs`, `BaseRLLossContext`.
  - Refactors GRPO/Oreal to the new API; 
  - `BaseRLLossContext` internalizes rollout importance-sampling handling.
- Worker/Trainer flow:
  - Workers and trainer now only interact with LossConfig and LossContext directly 
  - batching is done via `LossContext.build_batches`.
  - `compute_actor_logprobs` and `compute_ref_logprobs` return `List[torch.Tensor]` for clearer data flow.
  - SP handling is centralized within builders for consistency.

### Compatibility
- Public training CLI and typical user workflows remain unchanged; only internal wiring and APIs between workers/trainers and loss configs are updated.

### Testing
- Updated unit and integration tests across engines/models/losses to the new API.
- Verified identical reward curves , correct SP behavior for RL
image

### CODE REVIEW

File: xtuner/v1/rl/base/worker.py [hunk 1]
```diff
@@ -642,21 +666,31 @@ def fit(self, data_batches: list[WorkerInputItem], rollout_idx: int) -> WorkerLo
...
+            batches_advantages = advantages_list[i : i + iters_per_step]
+            batches_old_logprobs = old_logprobs_list[i : i + iters_per_step]
+            batches_ref_logprobs = ref_logprobs_list[i : i + iters_per_step] if ref_logprobs_list is not None else None
+            batches_rollout_is_weights = rollout_is_weights_list[i : i + iters_per_step]
+            batches_rollout_logprobs = rollout_logprobs_list[i : i + iters_per_step]
+
+            # LossContext = loss_cfg.loss_ctx_cls
+            # batches_loss_kwargs = LossContext.build_batches_loss_kwargs(batches_loss_ctx_input, loss_cfg)
+
+            batches_loss_ctx = loss_cfg.build_batches(
```
Comment (HIT-cwh) [inline xtuner/v1/rl/base/worker.py]: build_batches 这个 classmethod 放到 loss cfg 下，是怎么考虑的呢？我也没太想好

File: xtuner/v1/rl/base/worker.py [hunk 2]
```diff
@@ -506,89 +514,104 @@ def fit(self, data_batches: list[WorkerInputItem], rollout_idx: int) -> WorkerLo
...
 
-            if len(rollout_logprobs_list) > 0:
+            if sp_gathered_rollout_logprobs_list[i] is not None:
                 # calculate importance sampling weights
                 cu_seq_lens = seq_ctx_list[i].cu_seq_lens_q
                 num_tokens = cu_seq_lens[1:] - cu_seq_lens[:-1]
 
-                old_logprobs = loss_ctx_input.old_logprobs
+                old_logprobs = old_logprobs_list[i]
                 if self.sp_mesh and self.sp_mesh.size() > 1:
```
Comment (HIT-cwh) [inline xtuner/v1/rl/base/worker.py]: 这里我其实预想的是，fit 函数内最初把所有需要 sp split 的内容全部切分之后，整个过程尽量不要再出现额外的 sp split 和 gather，这样更清晰，例如 is 计算需要用到 sp gather 过的内容，就尽量在需要用到的函数内部完成 sp gather，因为这个函数需要完整的序列，这个是函数内部的需求而不是 fit 里的需求。

不过这个需要进一步考虑下。

**Comment by HIT-cwh** [comment]
https://github.com/InternLM/xtuner/blob/8f8bf16456c662c06847790ad76aee529c7eb7c7/xtuner/v1/rl/base/worker.py#L770

_train_one_step_sft 这里也需要对应修改

**Comment by HIT-cwh** [comment]
是不是类似 CELossContextInputItem 这样的 class 都不需要了

File: xtuner/v1/loss/base_loss_ctx.py [hunk 3]
```diff
@@ -96,6 +96,9 @@ class BaseLossConfig(BaseModel):
     def loss_ctx_cls(self) -> type["BaseLossContext"]:
         raise NotImplementedError
 
+    def build_batches(self, *args, **kwargs) -> List["BaseLossContext"]:
```
Comment (HIT-cwh) [inline xtuner/v1/loss/base_loss_ctx.py]: 直接用 list 就好，不需要再 from typing import List 这样来 import

File: xtuner/v1/rl/base/worker.py (hunk 1 above)
Comment (jayhenry) [reply xtuner/v1/rl/base/worker.py]: 这是为了和其他Cfg行为保持一致，比如 ModelCfg.build() 后生成 Model。
LossCtxCfg.build(args) 后也直接生成LossCtx，一些静态变量（如mode=chunk）在 LossCtxCfg中指定，一些动态变量（如sp_mesh）在build的入参args中指定。
这样使用者（如Trainer）只需要和 LossCtxCfg和LossCtx交互。

File: xtuner/v1/rl/base/worker.py (hunk 2 above)
Comment (jayhenry) [reply xtuner/v1/rl/base/worker.py]: loss_cfg应该只做build得到具体LossCtx对象的操作，实际功能应该由LossCtx对象来完成。
这里应该把IS相关计算都放到 LossCtx中。

**Comment by jayhenry** [comment]
> 是不是类似 CELossContextInputItem 这样的 class 都不需要了

Yes
